README.md
# OfflineContentGenerator
[![dependency status][1]][2] [![dev dependency status][3]][4]

Render service (Offline Content Generator) for MediaWiki.

Ties together the MediaWiki [Collection] extension with
the [mw-ocg-bundler] and [mw-ocg-latexer]/[mw-ocg-texter]/etc backends.

## Running a development server
See [wiki](https://wikitech.wikimedia.org/wiki/OCG#Installing_a_development_instance)
for instructions on how to configure a local [Collection] extension to point
to this server.  But you can run a local instance with
```sh
apt-get install redis-server # runs a local redis service on port 6379
cd <repodir> ; npm install
./mw-ocg-service.js # starts a front end ocg service on port 17080
```

You also need to checkout and install [mw-ocg-bundler] and one or more
backends ([mw-ocg-latexer], [mw-ocg-texter], etc).  Although you can
specify any path to these in you like, the default configuration
expects them to be at the same directory level as `mw-ocg-bundler`,
eg:
```
$OCG/mw-ocg-service (this package)
$OCG/mw-ocg-bundler
$OCG/mw-ocg-latexer
$OCG/mw-ocg-texter
```

If you have installed [VisualEditor], no additional configuration will
be necessary: OCG will use the VisualEditor configuration to find an
appropriate Parsoid service and prefix.

In the absence of VisualEditor, you will still need to install
[Parsoid], and then configure OCG to use it.  You will launch
`mw-ocg-service` as `./mw-ocg-service.js -c localsettings.js` and
create a `localsettings.js` file containing:
```javascript
// for mw-ocg-service
module.exports = function(config) {
  // change the port here if you are running parsoid on a different port
  config.backend.bundler.parsoid_api = "http://localhost:8000";
  // the prefix here should match $wgDBname in your LocalSettings.php
  config.backend.bundler.parsoid_prefix = "localhost";
}
```

Parsoid would in turn be configured with its own `localsettings.js`
containing:
```javascript
// for Parsoid
exports.setup = function( parsoidConfig ) {
  // first argument here should match $wgDBname in your LocalSettings.php
  parsoidConfig.setInterwiki( 'localhost', 'http://localhost/path/to/your/mediawiki/api.php' );
  // optional:
  parsoidConfig.serverPort = 8000;
};
```

## Installing a production server on Ubuntu

Ensure that you have a redis server installed somewhere.

```sh
cd <repodir> ; npm install
ln -s <repodir>/mw-collection-ocg.conf /etc/init
initctl reload-configuration
service mw-collection-ocg start
```

Create `/etc/mw-collection-ocg.js` to configure the render service.
See [the default configuration](./defaults.js) for more details.

## Binary node modules
The following node binary module is required:
* sqlite3 (for `bundler` and `latex_renderer`)

In addition, better performance is obtained if the following binary
module is installed:
* hiredis (for this service)

Be aware of these when deploying to a new node version or machine
architecture.  You may need to `npm rebuild <package name>`.

## Logging
This software uses the bunyan logging framework. By default the framework
only logs to the system console. To add additional log transports and make it
useful, in the config file add lines like:

```
var bsyslog = require('bunyan-syslog');
config.logging = {
  streams: [{
    stream: process.stdout
  }, {
    level: 'debug',
    type: 'raw',
    stream: bsyslog.createBunyanStream({
      type: 'sys',
      facility: bsyslog.local0,
      name: 'mw-ocg-service'
    })
  }, {
    type: 'raw',
    stream: require('gelf-stream').forBunyan( '..host..', PORT)
  }],
  serializers: ... /* optional */
}
```

## Maintenance
The health check ```command=health``` will return the number of objects
currently in the job queue. If the job queue is too long, the operator can
either wait for the queue to clear, or log into the redis server with the
rediscli and run ```DEL``` with the keyname given by ```JobQueue.name``` in
the returned health hash, equivilant to the value of ```config.redis.job_queue_name```

## License

(c) 2014 by Brad Jorsch, C. Scott Ananian, Matthew Walker, Max Seminik

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
http://www.gnu.org/copyleft/gpl.html

[Collection]:     https://www.mediawiki.org/wiki/Extension:Collection
[VisualEditor]:   https://www.mediawiki.org/wiki/Extension:VisualEditor
[Parsoid]:        https://www.mediawiki.org/wiki/Parsoid
[mw-ocg-bundler]: https://github.com/wikimedia/mediawiki-extensions-Collection-OfflineContentGenerator-bundler
[mw-ocg-latexer]: https://github.com/wikimedia/mediawiki-extensions-Collection-OfflineContentGenerator-latex_renderer
[mw-ocg-texter]:  https://github.com/wikimedia/mediawiki-extensions-Collection-OfflineContentGenerator-text_renderer

[1]: https://david-dm.org/wikimedia/mediawiki-extensions-Collection-OfflineContentGenerator.png
[2]: https://david-dm.org/wikimedia/mediawiki-extensions-Collection-OfflineContentGenerator
[3]: https://david-dm.org/wikimedia/mediawiki-extensions-Collection-OfflineContentGenerator/dev-status.png
[4]: https://david-dm.org/wikimedia/mediawiki-extensions-Collection-OfflineContentGenerator#info=devDependencies


CHANGELOG.md
# mw-ocg-service x.x.x (not yet released)

# mw-ocg-service 1.0.0 (2014-11-05)
* First release; corresponds to `mw-ocg-bundler` 1.2.0,
  `mw-ocg-latexer` 0.3.0 and `mw-ocg-texter` 0.3.0.


defaults.js
/**
 * Default configuration file for the Collection extension Offline Content
 * Generator. Local settings should be in /etc/mw-collection-ocg.js with
 * settings modified in a function like:
 *
 * module.exports = function(config) { config.foo = 'bar'; }
 */
module.exports = {
	/** Location of local configuration override(s) (if present) */
	"config": "/etc/mw-collection-ocg.js",

	/** Service management thread, coordinates (re)launching threads and initial global setup */
	"coordinator": {
		/** The number of frontend threads to spawn. At the moment we don't have good data on how
		 * many clients can be served via a thread in a production environment.
		 */
		"frontend_threads": 2,
		/** The number of backend threads to spawn. These are heavy rendering threads and so should
		 * be set to some ratio of CPU cores. If set to "auto" the coordinator will launch a thread
		 * per CPU core.
		 */
		"backend_threads": "auto",

		/** Public hostname of this instance for HTTP GET requests for locally stored content. */
		"hostname": null
	},
	/** Configuration for the frontend HTTP server thread. You can choose to serve
	 * content via a local socket, or an IP address. If both are null the server will
	 * bind to all IP addresses.
	 */
	"frontend": {
		"address": null,
		"socket": null,
		"port": 17080,
		/** Failed jobs won't be retried until afer the lockout time (in seconds).
		 * Should be <= garbage_collection.failed_job_lifetime
		 * A value of zero means that no time limit is enforced.
		 */
		"failed_job_lockout_time": 0
	},
	/** Configuration for the backend bundling & and rendering process threads. */
	"backend": {
		"bundler": {
			/** {int} Maximum time, in seconds, that the process will be allowed to execute.
			 * After the expiration time a SIGTERM will be issued. A value of zero means that
			 * no time limit is enforced.
			 */
			"max_execution_time": 0,
			"bin": "../mw-ocg-bundler/bin/mw-ocg-bundler",
			// turn off some mwlib compatibility features which bloat the bundle
			// set a maximum image size: 150dpi * 4" wide image.
			"additionalArgs": ['--no-compat', '--size=600'],

			"parsoid_api": "http://localhost/",
			"parsoid_prefix": "localhost"
		},
		"writers": {
			"rdf2latex": {
				/** {int} Maximum time, in seconds, that the process will be allowed to execute.
				 * After the expiration time a SIGTERM will be issued. A value of zero means that
				 * no time limit is enforced.
				 */
				"max_execution_time": 0,

				"bin": "../mw-ocg-latexer/bin/mw-ocg-latexer",
				"additionalArgs": [],
				"mimeType": "application/pdf",
				"extension": ".pdf"
			},
			"rdf2text": {
				/** {int} Maximum time, in seconds, that the process will be allowed to execute.
				 * After the expiration time a SIGTERM will be issued. A value of zero means that
				 * no time limit is enforced.
				 */
				"max_execution_time": 0,

				"bin": "../mw-ocg-texter/bin/mw-ocg-texter",
				"additionalArgs": [],
				"mimeType": "text/plain",
				"extension": ".txt"
			}
		},
		/* Compatibility aliases for the writer strings. */
		"writer_aliases": {
			// mwlib compatibility
			"rl": "rdf2latex"
		},

		/** {string} Working directory for the service. If null will be in the OS temp dir. */
		"temp_dir": null,
		/** {string} Directory for final rendered output. If null will be temp_dir/ocg-output */
		"output_dir": null,
		/** {string|null} Directory where failed jobs will be stored. If null will not be stored. */
		"post_mortem_dir": null
	},
	/** Redis is used in both the frontend and backend for queueing jobs and job
	 * metadata storage.
	 */
	"redis": {
		"host": "localhost",
		"port": 6379,
		"password": null,
		"retry_max_delay": 60000,

		"job_queue_name": "ocg_render_job_queue",
		"status_set_name": "ocg_job_status",

		/** {int} When the job queue is larger than this, new jobs will be rejected. A zero value
		 * means no limit to the number of jobs that can be in the queue.
		 */
		"max_job_queue_length": 0
	},
	/** Active metric reporting via the StatsD protocol. General health can be obtained by querying
	 * the frontend with a HTTP GET ?request=health query
	 */
	"reporting": {
		/** If true will send UDP packets to the StatsD server. */
		"enable": false,
		/** Hostname to send StatsD metrics to. */
		"statsd_server": "localhost",
		/** Port to send StatsD metrics to. */
		"statsd_port": 8125,
		/** The txstatsd daemon can have non standard behaviour. If you're running the
		 * ConfigurableCollector set this to true.
		 */
		"is_txstatsd": false,
		/** Prefix for all statistics generated by this application */
		"prefix": "ocg.pdf."
	},
	/** Bunyan logging configuration.  `streams` and `serializers` are the
	 * important properties here. */
	"logging": {
		/*
		streams: [
			{
				// Bare bones console output
				stream: process.stdout
			},
			{
				// Log to syslog on localhost
				level: 'debug',
				type: 'raw',
				stream: require('bunyan-syslog').createBunyanStream({
					type: 'sys',
					facility: 16,
					name: 'mw-ocg-service'
				})
			},
			{
				// log via gelf-stream
				type: 'raw',
				stream: require('gelf-stream').forBunyan( GELF_LOG_HOST, GELF_LOG_PORT )
			}
		],
		serializers: require('bunyan').stdSerializers
		*/
	},
	/** Garbage collection thread */
	"garbage_collection": {
		/** Seconds between garbage collection runs */
		every: 0.25 * 24 * 60 * 60,
		/** Lifetime, in seconds, of a job status object in redis */
		job_lifetime: 4 * 24 * 60 * 60,
		/** Lifetime, in seconds, of any successful job artifacts on the file system */
		job_file_lifetime: 4.25 * 24 * 60 * 60,
		/** Lifetime, in seconds, of a job status object that failed in redis */
		failed_job_lifetime: 24 * 60 * 60,
		/** Lifetime, in seconds, of an object in the temp file system. Must be longer
		 * than the longest expected job runtime. (We check ctime not atime)
		 */
		temp_file_lifetime: 0.25 * 24 * 60 * 60,
		/** Lifetime, in seconds, of an object in the post mortem directory. */
		postmortem_file_lifetime: 1 * 24 * 60 * 60
	}
};
